module_3/gift_certificates_system/repository/src/main/java/com/epam/esm/dao/FindAllDao.java

```diff
@@ -0,0 +1,75 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.exception.IncorrectArgumentException;
+import com.epam.esm.util.EsmPagination;
+import org.apache.commons.lang3.math.NumberUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.util.MessagePropertyKey.EXCEPTION_ESM_PAGINATION_PAGE_OUT_OF_RANGE;
+
+public abstract class FindAllDao<T> {
+    @PersistenceContext
+    protected EntityManager em;
+
+    @Autowired
+    protected CriteriaBuilder cb;
+
+    /**
+     * Find all entities with paginate.
+     *
+     * @param pagination Contain number of page and count elements on page.
+     * @param entity        Entity.
+     * @return Set of entities.
+     */
+    public Set<T> findAll(EsmPagination pagination, Class<T> entity) {
+        validatePaginationOrElseThrow(pagination, entity);
+
+        CriteriaQuery<T> cq = cb.createQuery(entity);
+        Root<T> root = cq.from(entity);
+        cq.select(root);
+
+        return executeQuery(cq, pagination);
+    }
+
+    protected void validatePaginationOrElseThrow(EsmPagination pagination, Class<T> entity) {
```

Comment:
From my point of view pagination validation can be performed in service layer

Reply:
For this reason, for validate the page number from the request need total number of available elements from the database (this operation is performed in the `dao` layer), it was decided to validate in the `dao` layer.